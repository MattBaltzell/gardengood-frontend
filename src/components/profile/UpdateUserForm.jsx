import React, { useContext, useState } from "react";
import { useNavigate, Navigate, Link } from "react-router-dom";
import "../form/Form.css";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import { MyTextInput } from "../form/Forms";
import UserContext from "../auth/UserContext";
import { v4 as uuid } from "uuid";

const UserUpdateForm = ({ update, toast }) => {
  // Pass the useFormik() hook initial form values, a validate function that will be called when
  // form values change or fields are blurred, and a submit function that will
  // be called when the form is submitted

  const navigate = useNavigate();
  const currUser = useContext(UserContext);
  const [submissionErrors, setSubmissionErrors] = useState([]);

  const INITIAL_STATE = {
    firstName: currUser.firstName,
    lastName: currUser.lastName,
    email: currUser.email,
    zipCode: currUser.zipCode,
  };

  return !currUser ? (
    <Navigate to="/" />
  ) : (
    <>
      <Link to={`/users/${currUser.username}`} className="breadcrumb">
        Back to{" "}
        {currUser.username[0].toUpperCase() + currUser.username.slice(1)}'s
        account
      </Link>
      <main className="Form">
        <h1>Update Account</h1>

        <Formik
          initialValues={{ ...INITIAL_STATE }}
          validationSchema={Yup.object({
            firstName: Yup.string()
              .max(25, "Must be 25 characters or less")
              .required("Required"),
            lastName: Yup.string()
              .max(30, "Must be 30 characters or less")
              .required("Required"),
            email: Yup.string()
              .email("Invalid email address")
              .required("Required"),
            zipCode: Yup.string()
              .length(5, "Must be exactly 5 characters")
              .required("Required"),
          })}
          onSubmit={(values) => {
            setTimeout(async () => {
              const username = currUser.username;
              const res = await update({ username, ...values });
              if (res.message) {
                res.message.forEach((m) => {
                  toast("error", m);
                  setSubmissionErrors((errors) => [...errors, m]);
                });
              } else {
                return navigate(`/users/${username}`);
              }
            }, 400);
          }}
        >
          <Form>
            <MyTextInput
              label="User Name"
              name="username"
              type="text"
              disabled
              value={currUser.username}
            />

            <MyTextInput
              label="Email"
              name="email"
              type="text"
              autoComplete="email"
            />
            <MyTextInput label="First Name" name="firstName" type="text" />
            <MyTextInput label="Last Name" name="lastName" type="text" />
            <MyTextInput
              label="Zip Code"
              name="zipCode"
              type="text"
              pattern="[0-9]{5}"
            />
            {submissionErrors
              ? submissionErrors.map((e) => (
                  <div
                    key={uuid()}
                    className="error center"
                    data-testid="required-error"
                  >
                    {e}
                  </div>
                ))
              : null}
            <button type="submit">Update Account</button>
          </Form>
        </Formik>
      </main>
    </>
  );
};

export default UserUpdateForm;
