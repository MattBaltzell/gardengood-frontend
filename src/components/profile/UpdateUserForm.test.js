import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import UpdateUserForm from "./UpdateUserForm";

import UserContext from "../auth/UserContext";

const user = {
  username: "testuser",
  email: "test@test.com",
  firstName: "firstname",
  lastName: "lastname",
  zipCode: "36830",
  joinAt: "2023-03-03 18:24:48.215773",
  isAdmin: false,
};

describe("User component tests", function () {
  it("renders without crashing", function () {
    render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <UpdateUserForm />
        </UserContext.Provider>
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <UpdateUserForm />
        </UserContext.Provider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("gives error if invalid input", async function () {
    render(
      <MemoryRouter>
        <UserContext.Provider value={user}>
          <UpdateUserForm />
        </UserContext.Provider>
      </MemoryRouter>
    );
    const lastNameInput = screen.getByDisplayValue("lastname");
    fireEvent.change(lastNameInput, { target: { value: "" } });
    const buttonEl = screen.getByRole("button");
    fireEvent.click(buttonEl);
    const requiredError = await screen.findByText("Required");
    expect(requiredError).toBeInTheDocument();
    const requiredErrors = await screen.findAllByText("Required");
    expect(requiredErrors.length).toEqual(1);
  });
});
