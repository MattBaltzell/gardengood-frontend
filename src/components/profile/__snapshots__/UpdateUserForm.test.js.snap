// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component tests matches snapshot 1`] = `
<DocumentFragment>
  <a
    class="breadcrumb"
    href="/users/testuser"
  >
    Back to Testuser's account
  </a>
  <main
    class="Form"
  >
    <h1>
      Update Account
    </h1>
    <form
      action="#"
    >
      <label
        for="username"
      >
        User Name
      </label>
      <input
        class="text-input"
        disabled=""
        name="username"
        type="text"
        value="testuser"
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        autocomplete="email"
        class="text-input"
        name="email"
        type="text"
        value="test@test.com"
      />
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        class="text-input"
        name="firstName"
        type="text"
        value="firstname"
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        class="text-input"
        name="lastName"
        type="text"
        value="lastname"
      />
      <label
        for="zipCode"
      >
        Zip Code
      </label>
      <input
        class="text-input"
        name="zipCode"
        pattern="[0-9]{5}"
        type="text"
        value="36830"
      />
      <button
        type="submit"
      >
        Update Account
      </button>
    </form>
  </main>
</DocumentFragment>
`;
